### PR TITLE
[SYCL] Avoid creating unused objects.

### PR DESCRIPTION
**Pull request by vladimirlaz**
Use initialization list to initialize Device and Context fields
of program_impl class. Otherwise objects are created with default
constructors first and then updated with proper values.

Signed-off-by: Vladimir Lazarev <vladimir.lazarev@intel.com>

### CODE REVIEW

File: sycl/test/regression/program-link.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  return 0;
+}
+
+device get_device(std::string type) {
+  try {
+    if (type == "cpu") {
+      return cpu_selector().select_device();
+    } else if (type == "gpu") {
```
Comment (bader) [inline sycl/test/regression/program-link.cpp]: Please, don't use `else` after `return`.

File: sycl/test/regression/program-link.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      return gpu_selector().select_device();
+    } else if (type == "acc") {
+      return accelerator_selector().select_device();
+    } else {
+      std::cout << "Bad type: " << type << std::endl;
```
Comment (bader) [inline sycl/test/regression/program-link.cpp]: What about `host` device?

Comment (bader) [inline sycl/test/regression/program-link.cpp]: I think we should throw an exception for "unknown" device type.

File: sycl/test/regression/program-link.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    vector_class<program> prgList;
+    prgList.push_back(get_compiled_program<struct kernel1>(cxt, dev));
+    prgList.push_back(get_compiled_program<struct kernel2>(cxt, dev));
+    program res{prgList};
+  }
+
+  {
+    try {
+      vector_class<program> prgList;
+      program res{prgList};
```
Comment (bader) [inline sycl/test/regression/program-link.cpp]: I'm not sure I understand how this code is supposed to validate that "unused objects are avoided". Could you clarify, please?

File: sycl/test/regression/program-link.cpp (hunk 1 above)
Comment (vladimirlaz) [reply sycl/test/regression/program-link.cpp]: will fix

File: sycl/test/regression/program-link.cpp (hunk 2 above)
Comment (vladimirlaz) [reply sycl/test/regression/program-link.cpp]: My initial assumption was that this code does not make sense for host device. But I do not find any blocker in SPEC. So I will add it.

Comment (vladimirlaz) [reply sycl/test/regression/program-link.cpp]: will do

File: sycl/test/regression/program-link.cpp (hunk 3 above)
Comment (vladimirlaz) [reply sycl/test/regression/program-link.cpp]: It does not check object creation. But it checks that impacted functions does not regress and produce expected results.

**Comment by vladimirlaz** [comment]
> Please fix commit message. `unsused` -> `unused`, I guess.

fixed

File: sycl/include/CL/sycl/detail/program_impl.hpp [hunk 4]
```diff
@@ -50,14 +50,9 @@ class program_impl {
   // caching resolution
   program_impl(vector_class<std::shared_ptr<program_impl>> ProgramList,
                string_class LinkOptions = "")
-      : State(program_state::linked), LinkOptions(LinkOptions),
+      : State(program_state::linked), Context(ProgramList.at(0)->Context),
+        Devices(ProgramList.at(0)->Devices), LinkOptions(LinkOptions),
         BuildOptions(LinkOptions) {
-    // Verify arguments
-    if (ProgramList.empty()) {
-      throw runtime_error("Non-empty vector of programs expected");
```
Comment (Fznamznon) [inline sycl/include/CL/sycl/detail/program_impl.hpp]: Why this check is removed?

Comment (vladimirlaz) [reply sycl/include/CL/sycl/detail/program_impl.hpp]: vector::at(0) is used instead. It does the check and generates std::out_of_range if vector is empty.
From my perspective it is more or less equal.

Comment (Fznamznon) [reply sycl/include/CL/sycl/detail/program_impl.hpp]: I think old exception described the problem better.

Comment (vladimirlaz) [reply sycl/include/CL/sycl/detail/program_impl.hpp]: but it does not allow to avoid unused object creation.

Comment (Fznamznon) [reply sycl/include/CL/sycl/detail/program_impl.hpp]: @romanovvlad , please provide your opinion on this.

**Comment by vladimirlaz** [comment]
Close to prepare general solution.

File: sycl/include/CL/sycl/detail/program_impl.hpp (hunk 4 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/program_impl.hpp]: I tend to agree with @Fznamznon . It's better to use sycl type of exception with more clear message here.

**Pull request closed without merging**